doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,271 @@
+If the security checks are not enabled then it is unapproved?
+
+Implementation options
+----------------------
+
+The above requirements indicate 2 options.
+
+### Option 1
+
+Dont allow ANY non approved algorithms and then a indicator is not required.
+
+- Pros: Simple
+- Cons: Problematic since we already have configurable options that are used for security checks etc.
+- Cons: We would need to return errors anywhere where an algorithm is not approved, which would cause compatibility issues
+
+### Preferred Option
+
+Add an indicator everywhere that it is required.
+
+- Pros: Flexible solution
+- Cons: Requires a lot more effort to add the indicator to all the required places.
+
+Note that in order for a service to be ‘fips approved’ the following requirements would need to be met.
+
+- Any algorithms come from FIPS provider.
+- A service is a series of one or more API calls that must all succeed
+- A extra API call is needed after the service succeeds, that should return 1 if the service is FIPS approved.
+
+Solutions for the preferred Option
+----------------------------------
+
+### Solution 1 (Using an indicator everywhere)
+
+Use a per thread global counter that is incremented when an algorithm is approved.  AWS/google have added this in places where a service is at the point of completing its output (e.g. digest_final). This design is complicated by the fact that a service may call another service (e.g HMAC using SHA256) that also wants to increment the approved counter. To get around this issue they have a second variable that is used for nesting. If the variable is non zero then the approved counter doesnt increment. This also allows non security relevant functions to not increment the approved count. Another variation of this would be to use flags instead of a counter.
+
+- Cons: At the fips provider level this would require some plumbing to go from the core to the fips provider, which seems overly complex.
+- Cons: The query can only be done after the output is set.
+- Cons: The indicator code would end up having to be set in different places depending on the algorithm after the output is finalized. This would be fairly messy as the point where it is called is set could be different for different algorithms.
+- Cons: The locking increment seems messy.
+
+### Proposed Solution (Using an indicator everywhere)
+
+Add a OSSL_PARAM getter to each provider algorithm context.
+By default if the getter is not handled then it would return not approved.
+
+- Pros: The code is easier to find since it is part of the get_ctx_params function.
+- Pros: The getter can be called at any point after the setting is done.
+
+Any fips algorithm that is approved would then need a setter that at a minimum contains code similar to the following
+
+``` C
+int ossl_xxx_fips_approved(void)
+{
+ifdef FIPS_MODULE
+    return 1; // conditional code would go here for each algorithm if required *
+else
+    return 0;
+endif
```

Comment:
Nothing requires adding the indicator params to the default provider, so I'd just make the `OSSL_FIPS_PARAM_APPROVED` param only work in the FIPS provider by wrapping its use in the `CTX_get_params()` function in `#ifdef FIPS_MODULE`. The `ossl_xxx_fips_approved()` function would then only have to be implemented in the FIPS provider.